### ISSUE TITLE
Suspecting potential racing condition on Ensemble input mapping

### ISSUE THREAD

**Issue by shixianc** [triton-inference-server/server#6334]
**Description**
We're suspecting race condition on how inputs/outputs get passed to other models in TritonServer model ensemble.
Note: all our models are using ````Triton Python Backend```

From the below config.pbtxt, basically within a request ```preprocessor``` model outputs ```input_ids + data_identifier_1```, and this ```data_identifier_1```  is passed to both ```predictor``` model and ```postprocessor``` model. However we observed in production, occasionally the ```data_identifier_1``` received in postprocessor does not match the ```predictions``` passed from ```predictor``` model. In other words, **postprocessor model received inputs from 2 different requests.** 

Though this is hypothetical and currently not able to reproduce locally, we want to check with Nvidia team here if you have seen similar issue. 

Minimal config.pbtxt:
```
name: "ensemble"
platform: "ensemble"
max_batch_size: 0
input [
  {
    name: "request"
    data_type:  TYPE_STRING
    dims: [1]
  }
]
output [
  {
    name: "response"
    data_type:  TYPE_STRING
    dims: [1]
  }
]

ensemble_scheduling {
  step [
    {
      model_name: "preprocessor"
      model_version: 1
      input_map {
        key: "request"
        value: "request"
      }
      output_map {
        key: "input_ids"
        value: "input_ids"
      }
      output_map {
        key: "data_identifier_1"
        value: "data_identifier_1"
      }
    },
    {
      model_name: "predictor"
      model_version: 1
      input_map {
        key: "input_ids"
        value: "input_ids"
      }
      input_map {
        key: "data_identifier_1"
        value: "data_identifier_1"
      }
      output_map {
        key: "PREDICTIONS"
        value: "PREDICTIONS"
      }
    },
    {
      model_name: "postprocessor"
      model_version: 1
      input_map {
        key: "PREDICTIONS"
        value: "PREDICTIONS"
      }
      input_map {
        key: "data_identifier_1"
        value: "data_identifier_1"
      }
      output_map {
        key: "response"
        value: "response"
      }      
    }  
  ]
}

````

**Triton Information**
What version of Triton are you using?
```nvcr.io/nvidia/tritonserver:22.01-pyt-python-py3```

Are you using the Triton container or did you build it yourself?
Use tritonserver 22.01 as base image, and then build our own image (just to add dependencies and libraries)

**To Reproduce**
Not able to reproduce - this is a hypothetical ask. 

Describe the models (framework, inputs, outputs), ideally include the model configuration file (if using an ensemble include the model configuration file for that as well).
Pytorch 1.8

**Expected behavior**
Expect the ensemble inputs/ouputs are passed in sequence within a single request.

**Comment by oandreeva-nv** [comment]
Hi @shixianc , thank you for reporting this. What version of Triton were you building?

**Comment by shixianc** [comment]
@oandreeva-nv  TritonServer we used is 2.18.0

**Comment by oandreeva-nv** [comment]
@shixianc, we've fixed some similar race conditions in 23.05 (v2.34.0): issue was raised [here](https://github.com/triton-inference-server/server/issues/5733) and the PR is [here](https://github.com/triton-inference-server/python_backend/pull/245)

**Comment by shixianc** [comment]
@oandreeva-nv Thanks the issue looks like to with GPU and cuda synchronization. But our use case is PythonBackend using AWS Inferentia. 
I'll see if there's anything similar in python backend though.

**Comment by oandreeva-nv** [comment]
Interesting, thank you for specifying AWS Inferentia, I wasn't aware of this. Let me know, if this issue still persists on later versions, and I can file a ticket for the team to investigate.

**Comment by dyastremsky** [comment]
Closing ticket due to inactivity. If you would like to reopen it for follow-up, let us know.

**Comment by shixianc** [comment]
To follow up on the issue, we made 2 changes and now the issue has been resolved: 
1. upgraded to TritonServer==23.08
2. Avoid passing any tensors from the first python model directly to the last python model, and instead "pass through" the predictor python model in the middle (i.e. each python model can only receive request from a single python model).

Not sure which of the above was the root cause though, but just want to post here for others facing same issue.
